feat: support dialog map mode

shiquda opened the pull request:
<!-- Template from https://github.com/kubevirt/kubevirt/blob/main/.github/PULL_REQUEST_TEMPLATE.md?-->
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/CherryHQ/cherry-studio/blob/main/CONTRIBUTING.md
-->

### What this PR does

Before this PR:

After this PR: 添加节点对话支持。

https://github.com/user-attachments/assets/f1313e1f-459a-44a1-80da-a5b93d691d77

由于是用Cursor糊的代码，可能还需要进一步打磨和debug

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->

https://github.com/CherryHQ/cherry-studio/discussions/4040

### Why we need it and why it was done in this way

The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Breaking changes

<!-- optional -->

If this PR introduces breaking changes, please describe the changes and the impact on users.

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature.

### Release note

<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
3. If no release note is required, just write "NONE".
-->

```release-note

```

## Summary by Sourcery

Introduce a dialog map feature to visualize conversation history as a branching tree, allowing users to explore and manage different conversational paths.

New Features:
- Add a dialog map view to visualize conversation branches.
- Allow selecting a specific path in the conversation tree.
- Enable creating new conversation branches from existing messages.
- Add a button in the header to open the dialog map view.

Chores:
- Define data types (`DialogMap`, `DialogMapNode`) and database schema for storing dialog map data.
- Implement a service (`DialogMapService`) to handle dialog map creation, updates, and data retrieval.
- Add an event (`MESSAGES_UPDATED`) for message updates related to the map.
- Include translations for the new feature.new translations for the dialog map feature.

## Summary by Sourcery

Introduce a dialog map feature to visualize and manage conversation history as a branching tree.

New Features:
- Add a dialog map view, accessible via a new header button, to visualize conversation branches.
- Allow selecting a specific path in the conversation tree.
- Enable creating new conversation branches from existing messages.
- Support deleting conversation branches (nodes and their descendants).
- Apply the selected path in the map to update the main conversation view.
- Allow navigating from a map node back to the corresponding message in the main chat view (if the node is part of the current path).

Chores:
- Define data structures (`DialogMap`, `DialogMapNode`) and update the database schema to store dialog map data.
- Add translations for the dialog map feature.
- Add a `MESSAGES_UPDATED` event for message updates related to the map.
- Implement a `DialogMapService` to handle dialog map creation, updates, path selection, and data retrieval.
- Add utility functions for dialog map data processing and layout calculations for the visualization.

shiquda committed:
fix(OpenAIProvider): ensure messages.content of the request is string

shiquda committed:
feat: support dialog map mode

shiquda committed:
fix: remove unused db store

shiquda committed:
feat: refine dialog map style

shiquda committed:
feat: delete dialog map node

shiquda committed:
feat: refine tree style

pilgrimlyieu commented:
这个功能似乎有了：
- #3772 

首先将「对话导航按钮」改为「上下按钮」，然后最后一个就是「聊天历史」。

不过这个似乎是唯一的入口，但是我更常用的是「对话锚点」，所以要打开也是挺麻烦的。

shiquda committed:
feat: locate message in dialog map

shiquda commented:
> 这个功能似乎有过一个实现：
> 
> * [feat：增加聊天记录流程图，方便查看 #3772](https://github.com/CherryHQ/cherry-studio/pull/3772)
> 
> 首先将「对话导航按钮」改为「上下按钮」，然后最后一个就是「聊天历史」。
> 
> 不过这个似乎是唯一的入口，但是我更常用的是「对话锚点」，所以要打开也是挺麻烦的。

对话历史只支持线性的对话显示，我现在做的是实现对话分支功能，差别还是比较大的，不过流程图部分的渲染参考了聊天历史的代码

shiquda commented:
@sourcery-ai review

shiquda committed:
refactor: simplify DialogMap component data flow and state management

shiquda committed:
fix: improve border color referencing in CustomNodeContainer styled component

shiquda commented:
@kangfenmao 看看这个idea怎么样？

shiquda committed:
feat: update i18n for dialog map

shiquda committed:
refactor(dialogMapUtils): extract recursive child position adjustment into helper function

shiquda committed:
fix(DialogMap): import bezier type

shiquda committed:
fix: remove unnecessary title of node card

shiquda commented:
![PixPin_2025-04-28_22-45-20](https://github.com/user-attachments/assets/456f7f2f-ca22-4a24-bef7-a843bb134a03)

现在是一个勉强能用的状态，还有一些bug没调，自己build先用上了😄。现在有一个的问题，如果对话很长的话，怎么样让用户快速找到需要的节点，而不显得太长或者太杂乱？看看大家有没有什么想法，UI怎么调整一下，比如可以折叠分支之类的

![PixPin_2025-04-28_22-45-20](https://github.com/user-attachments/assets/474d83ca-b28d-48c8-85eb-8725272307db)

shiquda committed:
Merge branch 'upstream/main' into feature/chat-map

CreatorZZY commented:
> ![PixPin_2025-04-28_22-45-20](https://private-user-images.githubusercontent.com/104683621/438254715-456f7f2f-ca22-4a24-bef7-a843bb134a03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU5OTk2NzgsIm5iZiI6MTc0NTk5OTM3OCwicGF0aCI6Ii8xMDQ2ODM2MjEvNDM4MjU0NzE1LTQ1NmY3ZjJmLWNhMjItNGEyNC1iZWY3LWE4NDNiYjEzNGEwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQwNzQ5MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMTMzMWRlYmRkNWE2ZTk5ZjkzYTYyYWRiODBiNjBhMGJhNmJjMDJlNTVjNjBiNmZkOTQ4MDUzN2YyMjMyNGY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tBIKNjhm7PHS3L6OmoqM3Va4yM7dqF75lcgodx_kGZM)
> 
> 现在是一个勉强能用的状态，还有一些bug没调，自己build先用上了😄。现在有一个的问题，如果对话很长的话，怎么样让用户快速找到需要的节点，而不显得太长或者太杂乱？看看大家有没有什么想法，UI怎么调整一下，比如可以折叠分支之类的

可以自定义Compact显示模式：
1. 可以折叠模型的回答
2. 只显示提问
3. 箭头的长度也缩短。
4. 模型的Logo缩小、直接与提问同一行。
5. 减少框内部的padding。

shiquda committed:
feat(DialogMap): refine map ui

- Add node collapse/expand functionality
- Optimize node merging display logic
- Adjust styling and layout

1411124232 commented:
还不错 还需要添加触控板双指拖移
src/renderer/src/pages/home/Messages/DialogMap.tsx文件
 panOnScroll={false}改 panOnScroll={true}

shiquda commented:
![PixPin_2025-04-30_17-16-20](https://github.com/user-attachments/assets/79c76252-84b2-40a8-999b-fe3bdb8210ec)

shiquda committed:
feat: add support for trackpad two-finger drag

1411124232 commented:
有BUG 你新加的 
点击卡片会自动收缩
第二个模型头像没有

1411124232 commented:
还需要增加一个记录折叠状态 现在折叠过 每次重新点进去 都是全折叠状态

shiquda committed:
fix: keep node collapse status correctly

shiquda committed:
refactor(dialogMap): create DialogMapNode file

shiquda committed:
feat(dialogMap): disable zoom on double click

shiquda committed:
fix(dialogMap): fix model avatar

shiquda committed:
feat(dialogMap): i18n

shiquda committed:
refactor(DialogMapNode): Optimize the structure of Tooltip and Dropdown, improve the display logic of operation buttons

shiquda commented:
上面提到的应该都修了

1411124232 commented:
还有问题 修改后 优点烧CPU

1411124232 commented:
问题找到了，是这个src/renderer/src/utils/dialogMapUtils.ts文件的问题 当大于2个图时 就开始烧cpu了
需要禁用这个动画animated: false，不然大于2个图就开始烧cpu

1411124232 commented:
<img width="143" alt="image" src="https://github.com/user-attachments/assets/9494d96f-edae-47d9-9620-de56537c0daf" />
悬停时显示全白色，这个文件src/renderer/src/pages/home/Messages/DialogMapNode.tsx
 background-color: var(--color-primary-dark);改成background-color: var(--color-primary);

demmosee commented:
非常希望能增加一个功能，就是可以选择任意的聊天点，一起发送到一个新话题中（就是可以选择所有对话中，最重要，最高质量的问答一起生成一个新话题），这样可以高效避免太长的上下文和不重要的信息对AI思考进行的干扰。

demmosee commented:
> 非常希望能增加一个功能，就是可以选择任意的聊天点，一起发送到一个新话题中（就是可以选择所有问答中，最重要，最高质量的问答一起生成一个新话题），这样可以高效避免太长的上下文和太多不重要的重复信息对AI思考进行的干扰。

shiquda commented:
这个PR可能需要其他dev的帮助，我最近没有太多精力来开发这个功能，而且已有的代码大多是Cursor写的，估计存在不少坑。看看有没有对这个功能感兴趣的dev来帮助开发

Pleasurecruise committed:
Merge branch 'main' into feature/chat-map

Pleasurecruise committed:
fix: message type error

CreatorZZY commented:
![image](https://github.com/user-attachments/assets/cf243c0b-e0ca-4861-9c50-954c9d767806)

并没有绘制出新的branch。而是直接添加到最末尾的地方。

shiquda commented:
> ![image](https://private-user-images.githubusercontent.com/38124587/440190770-cf243c0b-e0ca-4861-9c50-954c9d767806.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzNjgyNjAsIm5iZiI6MTc0NjM2Nzk2MCwicGF0aCI6Ii8zODEyNDU4Ny80NDAxOTA3NzAtY2YyNDNjMGItZTBjYS00ODYxLTljNTAtOTU0YzlkNzY3ODA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDE0MTI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1OThkNzRiODNiOWJlN2MzN2Y3MGE5NjdkODljZjE5OTViZTBmOGM3YzMzODU1MmU4ODY3NDAzZjQwYTEzMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.v_PUo7_-FulFHp7GX0TaBCN6y6FNLg6jxaZnQPDlVZ0)
> 
> 并没有绘制出新的branch。而是直接添加到最末尾的地方。

如果只有一个子节点就按照正常垂直排列，这是合理的

CreatorZZY commented:
> 如果只有一个子节点就按照正常垂直排列，这是合理的，如果在已有子节点的情况下在父节点点加号，才会有新的分支

![image](https://github.com/user-attachments/assets/4ee85208-fa7c-489d-ac33-0f4ff9a5ace9)

shiquda commented:
> > 如果只有一个子节点就按照正常垂直排列，这是合理的，如果在已有子节点的情况下在父节点点加号，才会有新的分支
> 
> ![image](https://private-user-images.githubusercontent.com/38124587/440191448-4ee85208-fa7c-489d-ac33-0f4ff9a5ace9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzNjg4NDYsIm5iZiI6MTc0NjM2ODU0NiwicGF0aCI6Ii8zODEyNDU4Ny80NDAxOTE0NDgtNGVlODUyMDgtZmE3Yy00ODlkLWFjMzMtMGY0ZmY5YTVhY2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDE0MjIyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0MjQ5NzgwYjk4M2MzMjI1NDE1Y2RjM2IxNmQ4ZTQ0OTVhYTU1YzVmNGZjOTFhYzY1MTRiN2E4NWRmN2MwYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8CPdnc5a6NDyUOA5NQLVh45v01zTipBO8ISB3ICyecU)
> 
> 按道理来说，有三个节点的情况下，在第二个节点点击添加分支的时候，应该按照我画的路线进行才对。

你的意思是不是相对位置的问题？比如说现在是

```
1
|
2 
| 
3 

```

在2添加分支，预期是这样

```
1
|
2 - 
|    |
3   4
     |
     5
```

但实际可能是这样

```
1
|
2 - 
|    |
4   3
|
5
```

如果是的话，确实存在这个问题

CreatorZZY commented:
我也clone了一份实现类似的功能，我的思路简单来说是这样：
1. 每次添加分支时把分支赋予branch-id为branch-{timestamp}。
2. 绘图的时候对同一个topic下的branch-{timestamp}进行排序，按顺序设置横坐标。
3. 新建或选择branch后，每一个新的Message都会带有对应branch-id、branch父节点的node-id信息，以方便回溯整个path。
4. 新建或选择branch后，原有的ScrollContainer中的非branch-id被隐藏不显示、调用大模型api时不被包括作为背景。

CreatorZZY commented:
> > > 如果只有一个子节点就按照正常垂直排列，这是合理的，如果在已有子节点的情况下在父节点点加号，才会有新的分支
> > 
> > 
> > ![image](https://private-user-images.githubusercontent.com/38124587/440191448-4ee85208-fa7c-489d-ac33-0f4ff9a5ace9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYzNjg4NDYsIm5iZiI6MTc0NjM2ODU0NiwicGF0aCI6Ii8zODEyNDU4Ny80NDAxOTE0NDgtNGVlODUyMDgtZmE3Yy00ODlkLWFjMzMtMGY0ZmY5YTVhY2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA0VDE0MjIyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0MjQ5NzgwYjk4M2MzMjI1NDE1Y2RjM2IxNmQ4ZTQ0OTVhYTU1YzVmNGZjOTFhYzY1MTRiN2E4NWRmN2MwYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8CPdnc5a6NDyUOA5NQLVh45v01zTipBO8ISB3ICyecU)
> > 按道理来说，有三个节点的情况下，在第二个节点点击添加分支的时候，应该按照我画的路线进行才对。
> 
> 你的意思是不是相对位置的问题？比如说现在是
> 
> ```
> 1
> |
> 2 
> | 
> 3 
> ```
> 
> 在2添加分支，预期是这样
> 
> ```
> 1
> |
> 2 - 
> |    |
> 3   4
>      |
>      5
> ```
> 
> 但实际可能是这样
> 
> ```
> 1
> |
> 2 - 
> |    |
> 4   3
> |
> 5
> ```
> 
> 如果是的话，确实存在这个问题

就是这个意思

shiquda commented:
> 我也clone了一份实现类似的功能，我的思路简单来说是这样：
> 
> 1. 每次添加分支时把分支赋予branch-id为branch-{timestamp}，默认无分支的情况下branch-id为undefined。
> 2. 绘图的时候对同一个topic下的branch-{timestamp}进行排序，按顺序设置横坐标。遇到undefined的时候默认不调整横坐标。
> 3. 新建或选择branch后，每一个新的Message都会带有对应branch-id、branch父节点的node-id信息，以方便回溯整个path。
> 4. 新建或选择branch后，原有的ScrollContainer中的非branch-id被隐藏不显示、调用大模型api时不被包括作为背景。

这样似乎会简单一些，看了一下你分支的type定义改动比较少

shiquda commented:
BTW，message模块是不是重构了？我看我这个分支的历史记录都变成空白的了，新对话Map好像也无法正确识别对话内容

CreatorZZY commented:
> > 我也clone了一份实现类似的功能，我的思路简单来说是这样：
> > 
> > 1. 每次添加分支时把分支赋予branch-id为branch-{timestamp}，默认无分支的情况下branch-id为undefined。
> > 2. 绘图的时候对同一个topic下的branch-{timestamp}进行排序，按顺序设置横坐标。遇到undefined的时候默认不调整横坐标。
> > 3. 新建或选择branch后，每一个新的Message都会带有对应branch-id、branch父节点的node-id信息，以方便回溯整个path。
> > 4. 新建或选择branch后，原有的ScrollContainer中的非branch-id被隐藏不显示、调用大模型api时不被包括作为背景。
> 
> 这样似乎会简单一些，看了一下你分支的type定义改动比较少

那是因为我还没改到Message的保存部分，现在是在探索到底咋为新建/选择branch之后所创建的message添加branchid属性。

DeJeune commented:
不好添加的话，直接拓展消息块数据结构，只要做好旧数据到新数据的转换就可以了

Pleasurecruise commented:
> message模块是不是重构了

我合并了一下主分支 粗略处理了一下类型报错 具体逻辑还没有看

toddyoe commented:
希望话题列表也以树形结构展示，把分支话题纳入管理，相关issue：#5477
